### ISSUE TITLE
Error when using amp: expected scalar type float but found c10::Half

### ISSUE THREAD

**Issue by Cc-Hy** [pytorch/pytorch#63733]
## 🐛 Bug
when using amp, the program failed with ERROR 
"Exception has occurred: RuntimeError,
expected scalar type float but found c10::Half"

## To Reproduce
HERE IS THE CODE:
1. in train.py
`

        model = MONO_MERGE()
        model.model_2d = build_network(model_cfg=cfg.MODEL2D, num_class=len(cfg.CLASS_NAMES), dataset=train_set)
        model.model_3d = build_network(model_cfg=cfg.MODEL3D, num_class=len(cfg.CLASS_NAMES), dataset=train_set)

        if args.sync_bn:
            model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model)
        model.cuda()

        optimizer = build_optimizer_my(model.model_2d, cfg.OPTIMIZATION)

        model.model_2d.train()  # before wrap to DistributedDataParallel to support fixed some parameters

        if dist_train:
            model = nn.parallel.DistributedDataParallel(model, device_ids=[cfg.LOCAL_RANK % torch.cuda.device_count()], find_unused_parameters=True)

        ......

        optimizer.zero_grad()
        with amp.autocast():
            loss, tb_dict, disp_dict = model_func(model, batch)

        scaler.scale(loss).backward()

        clip_grad_norm_(
            model.module.model_2d.parameters() if isinstance(model, torch.nn.parallel.DistributedDataParallel) else model.model_2d.parameters(),
            optim_cfg.GRAD_NORM_CLIP
        )
        
        scaler.step(optimizer)
        scaler.update()
`

2.in the model (where error occurs)

`

    def transform_grid(self, voxel_grid, grid_to_lidar, lidar_to_cam, cam_to_img):

        r"""Transforms voxel sampling grid into frustum sampling grid
        Args:
            grid: (B, X, Y, Z, 3), Voxel sampling grid
            grid_to_lidar: (4, 4), Voxel grid to LiDAR unprojection matrix
            lidar_to_cam: (B, 4, 4), LiDAR to camera frame transformation
            cam_to_img: (B, 3, 4), Camera projection matrix
        Returns:
            frustum_grid: (B, X, Y, Z, 3), Frustum sampling grid
        """
        B = lidar_to_cam.shape[0]
        # Create transformation matricies
        V_G = grid_to_lidar  # Voxel Grid -> LiDAR (4, 4)
        C_V = lidar_to_cam  # LiDAR -> Camera (B, 4, 4)
        I_C = cam_to_img  # Camera -> Image (B, 3, 4)
        trans = C_V @ V_G

        # Reshape to match dimensions
        trans = trans.reshape(B, 1, 1, 4, 4)
        voxel_grid = voxel_grid.repeat_interleave(repeats=B, dim=0)

        # Transform to camera frame
        camera_grid = transform_points(trans_01=trans, points_1=voxel_grid)    **ERROR HERE**`

ERROR:
  Exception has occurred: RuntimeError
expected scalar type float but found c10::Half
  File ".../OpenPCDet/pcdet/models/backbones_3d/vfe/image_vfe_modules/f2v/frustum_grid_generator.py", line 94, in transform_grid
    camera_grid = transform_points(trans_01=trans, points_1=voxel_grid)
  File ".../OpenPCDet/pcdet/models/backbones_3d/vfe/image_vfe_modules/f2v/frustum_grid_generator.py", line 119, in forward
    frustum_grid = self.transform_grid(voxel_grid=self.voxel_grid.to(lidar_to_cam.device),
3. this is the function
`

    def transform_points(trans_01: torch.Tensor, points_1: torch.Tensor) -> torch.Tensor:
      r"""Function that applies transformations to a set of points.

    Args:
        trans_01 (torch.Tensor): tensor for transformations of shape
          :math:`(B, D+1, D+1)`.
        points_1 (torch.Tensor): tensor of points of shape :math:`(B, N, D)`.
    Returns:
        torch.Tensor: tensor of N-dimensional points.

    Shape:
        - Output: :math:`(B, N, D)`

    Examples:

        >>> points_1 = torch.rand(2, 4, 3)  # BxNx3
        >>> trans_01 = torch.eye(4).view(1, 4, 4)  # Bx4x4
        >>> points_0 = transform_points(trans_01, points_1)  # BxNx3
    """
    check_is_tensor(trans_01)
    check_is_tensor(points_1)
    if not trans_01.shape[0] == points_1.shape[0] and trans_01.shape[0] != 1:
        raise ValueError("Input batch size must be the same for both tensors or 1")
    if not trans_01.shape[-1] == (points_1.shape[-1] + 1):
        raise ValueError("Last input dimensions must differ by one unit")

    # We reshape to BxNxD in case we get more dimensions, e.g., MxBxNxD
    shape_inp = list(points_1.shape)
    points_1 = points_1.reshape(-1, points_1.shape[-2], points_1.shape[-1])
    trans_01 = trans_01.reshape(-1, trans_01.shape[-2], trans_01.shape[-1])
    # We expand trans_01 to match the dimensions needed for bmm
    trans_01 = torch.repeat_interleave(trans_01, repeats=points_1.shape[0] // trans_01.shape[0], dim=0)
    # to homogeneous
    points_1_h = convert_points_to_homogeneous(points_1)  # BxNxD+1
    # transform coordinates
    points_0_h = torch.bmm(points_1_h, trans_01.permute(0, 2, 1))
    points_0_h = torch.squeeze(points_0_h, dim=-1)
    # to euclidean
    points_0 = convert_points_from_homogeneous(points_0_h)  # BxNxD
    # reshape to the input shape
    shape_inp[-2] = points_0.shape[-2]
    shape_inp[-1] = points_0.shape[-1]
    points_0 = points_0.reshape(shape_inp)
    return points_0
`

## Environment

 - PyTorch Version: 1.7.1
 - OS: Linux
 - Python version: 3.8
 - CUDA/cuDNN version: 11.0
![QQ截图20210822212909](https://user-images.githubusercontent.com/82150240/130357200-d6715450-5fcf-4fd9-81bb-fd0dd73a9715.png)

cc @ngimel @mcarilli @ptrblck
